text/0000-introduce-a-new-signing-mechanism.md

```diff
@@ -0,0 +1,558 @@
+
+In Phase 1, signing will be opt-in, meaning that signatures will require `build --sign` and verification will require `build install --verify-signatures` (or equivalent `gem signatures` commands). But by Phase 3, signing has become default in `build` and verifying is default in `install`. This guide explanation will assume that we have reached Phase 3, as this is the most complete scenario.
+
+### Signing a gem at build time
+
+Gems are automatically signed at build time using the `gem build` command. For example, to build and sign `foo`:
+
+```console
+$ gem build foo
+Sign your gem by selecting an identity provider in your browser. Press Ctrl-C to abort.
+Gem `foo` signed as person@example.com
+```
+
+The `gem build` command opens a browser for the signature authentication flow (see [_Signature authentication flow_](#signature-authentication-flow)).
+
+Before you build the gem, you must ensure that your email address is listed in the gemspec's [`email`](https://guides.rubygems.org/specification-reference/#email) field.
+
+You may also use `gem build --sign` as in previous versions of RubyGems, but this is no longer necessary to trigger the signature process.
+
+If you do not wish to sign a gem when building, you can use the `--no-sign` parameter. For example, to build `foo` without signing it:
+
+```console
+$ gem build --no-sign foo
+WARN: gem `foo` was not signed. Clients > 9.9.99 will not install your gem by default. You can sign the gem with the `gem signatures` command.
+```
+
+You may then perform signing separately, as described in the section [_Signing a gem after build time_](#signing-a-gem-after-build-time). This functionality is intended for maintainers of multiple related gems, who may wish to build a number of gems and then sign them in a single operation.
+
+You should _always_ sign your gems, whether at build time or afterwards. Unsigned gems cannot be verified by signature, presenting a higher security risk to end users. For this reason, unsigned gems are treated as an error at installation by default.
+
+#### Signature authentication flow
+
+As part of the `gem build` command, the gem client will open the sigstore login page using your system web browser. To sign a gem, you must choose one of the identity providers that sigstore supports.
+
+<img width="60%" alt="sigstore sign-in" src="https://user-images.githubusercontent.com/33674553/151442880-563d3eaf-e965-4d98-88c9-b8c0dc1cbaed.png">
+If you are not already signed in with your selected provider, you will need to log in.
+
+<img width="60%" alt="Sign in to GitHub" src="https://user-images.githubusercontent.com/33674553/151442855-bc68dbcd-757d-40be-b38b-88d2be61e386.png">
+
+You then need to grant sigstore permission to request your email address from your provider account. This step only occurs on your first attempt to sign a gem using this provider. Only provider-verified email addresses may be used to sign gems.
```

Comment:
Just to confirm my reading here: the OIDC provider verified email here is **not** being checked against one (or more) valid emails for the RubyGems user identity?

In other words: is there any formal relationship between the signing email (on sigstore's side) and the email(s) registered to a RubyGems account?

Reply:
> In other words: is there any formal relationship between the signing email (on sigstore's side) and the email(s) registered to a RubyGems account?

In this design, no. We key instead on the `email` field in the gemspec, since that file is controlled at build time. Unfortunately there's no fixed relationship between `email` and the set of rubygems.org owners of a gem.

Where we think the connection back to owners will be possible is if we add push events to the transparency log. Put another way: authors would publish signature log entries, but rubygems.org would publish the push event log entries. In the latter case rubygems.org can annotate which owner made the push.

Reply:
Adding to the above, there's a security advantage in _not_ tying signatures to the owner emails as verified by rubygems.org. It creates a separate account, so that attackers need to achieve _two_ account takeovers to publish a signed but malicious gem. One for a rubygems account, one for an identity provider account.

Reply:
Thanks for the clarification! That makes sense.

Reply:
> Adding to the above, there's a security advantage in _not_ tying signatures to the owner emails as verified by rubygems.org. It means that attackers need to achieve _two_ account takeovers to publish a signed but malicious gem. One for a rubygems account, one for an identity provider account.

@jchestershopify I did not understand this point. I don't use gems so my understanding might be limited. Request you to please clarify. 
Let us assume that an attacker has taken over the account of a gem owner. Now, is it not possible for the attacker to do the following:
1. Modify the gemspec file to add a new `email` (or replace existing `email`) which is the attacker's email address
2. Sign using the newly added attacker `email` in the gemspec
3. Push the gem to rubygems.org using the owner's compromised account

Now, during `gem install`, the signature verification should work fine, since the gem is signed using an `email` associated with the gemspec.

Reply:
`email` in `gemspec` is in there just for informational purpose. It doesn't grant access for publishing of gem. Adding another user to be able to push the gem is done via `gem owner` command (https://guides.rubygems.org/command-reference/#gem-owner).

Reply:
> Adding to the above, there's a security advantage in not tying signatures to the owner emails

I don't believe so. If you have gem push rights, you can push a gemspec with any email addresses you want in it, of course. 

I think better integration between `gem owner` and the signature source of identity might actually be important for gem maintainer ergonomics and increasing ease of use and uptake here. As it is, a collaboratively-team-maintained gem will now have two pieces of metadata it needs to keep in sync, the gemspec email addreses (which were *previously* for informational purposes only, but now are important for verifying signed trust), as well as the `gem owner` system (in which rubygems has been gradually making it *harder* to find email addressees associated with gem ownership... which seems at cross-purposes here, where the proposal will put individual email addresess front and center again). 

Even if some changes to rubygems might have to be considered to allow leaning harder and/or integrating `gem owner` identity here, I think it might be worth it, for ergonomics and comprehensibilty of the system, rather than bifurcating into two separate lists of ownership, which may or may not match, are used for different things, one of which is built on exposing email addresses and one of which is trying to hide them, etc.

Reply:
> gem owner/gem push rights are still complete keys to the kingdom

Yes.

Reply:
>>Adding to the above, there's a security advantage in not tying signatures to the owner emails

>I don't believe so. If you have gem push rights, you can push a gemspec with any email addresses you want in it, of course. And therefore, I believe, easily add a new email address to be used for this proposed signing mechanism to the gemspec too. gem owner/gem push rights are still complete keys to the kingdom. (Am I missing something?)

Security mechanisms offer a mix of prevention, detection, and recovery from compromises. Gem signing would not _prevent_ a compromised owner account from being used to push a gem. But it means that the attacker is easier to _detect_, because unless they also control the email address previously used, they need to introduce a new email address. This increases the chances of detecting an anomaly. Recovery is made easier because of the trusted public log. It’s not as good as prevention (for which we are championing MFA), but it’s an improvement nevertheless.

>I think better integration between gem owner and the signature source of identity might actually be important for gem maintainer ergonomics and increasing ease of use and uptake here.

As you said, we should not create more ways for gem maintainers to screw up w.r.t. email addresses. Do you have any proposals for improving ergonomics, and to reduce the room for error?

Reply:
If gem authors are given extra work to authenticate their identity, then there would be the expectation that the authentication is used for authorization to make sure only gem owners are allowed to publish a package.

In this case, that means authorizing that the gem is signed by a gem owner, which means that the initial gem owner and changes to gem owners need to be signed and included in the transparency log such that they can be verified.

Currently, rubygems.org is the source of truth for the owners of a gem, so it needs to sign that initial gem owner list.  Further changes should be signed by one of the gem owners.  That way, the gem verification can check that a gem package is signed by one of the gem owners.

If the same were attempted to be done with fields on the gemspec, then we would need to verify against the list of owners in a previously verified version of the gem, not the gem being verified.  It would also mean that a gem would need to be published to change the list of owners.  In contrast, keeping gem ownership changes as separate events, that just need extra authentication, helps keep things working as ruby developers currently expect.

Reply:
> If gem authors are given extra work to authenticate their identity, then there would be the expectation that the authentication is used for authorization to make sure only gem owners are allowed to publish a package.

We see signing ("I attest this build") and pushing as separate operations. In the original design signing is entirely independent of rubygems.org; in the alternative it gets tied in to rubygems.org identities. But it's still a logically distinct operation.

> Currently, rubygems.org is the source of truth for the owners of a gem, so it needs to sign that initial gem owner list. Further changes should be signed by one of the gem owners. That way, the gem verification can check that a gem package is signed by one of the gem owners.

The idea of having existing owners sign new owners is an interesting one. I think we can reuse that when we revisit the topic of further log entry types in future.

Reply:
Signing and pushing are distinct actions, but the ability to do both should still be needed to publish a gem that will be accepted by a rubygems client for installation, such that they aren't really logically distinct.  If you mean that the rubygems.org shouldn't need reject the publishing of a gem without a valid signature, then that's correct that it shouldn't be needed for security reasons (it would just keep invalid data out of its database).

In the alternative where rubygems.org verifies the identity, then it should be possible to make the gem signing seamless from the perspective of the gem author, since they would otherwise have their identity verified twice in a row (for signing then pushing).  Since there is no security benefit from authenticating twice when these are done together, there should not be any extra work.  The seemless experience also means that there wouldn't be a need to opt-in to adopt this feature.  The only difference would be that the rubygems authentication for publishing would get persisted as a signature in the gem and events in the transparency log for verification.

Reply:
I agree that signing and pushing are distinct and that in the long run both should be verified (we touched on it in "Other kinds of log entries"). But we considered it out of scope for this RFC. We're trying to boil the tea kettle first.